### ISSUE TITLE
Untyped parameter in Subscribe method of ILifecycleObservable API

### ISSUE THREAD

**Issue by dVakulen** [dotnet/orleans#3802]
Method  [`Subscribe(int stage, ILifecycleObserver observer)`](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core.Abstractions/Lifecycle/ILifecycleObservable.cs#L19) implicitly requires specific enum value, while having `int` stage parameter, thus resulting in enum casting which leads to lose of the statically typed parameter benefits.  
### Downsides of untyped parameter usage: 
- Introduces need in [cast helper extension class](https://github.com/dotnet/orleans/blob/master/src/Orleans.Runtime/Lifecycle/SiloLifecycleExtensions.cs#L7) which brings additional entity with error-prone implementation to support.
- There's no restriction from using API while avoiding cast extensions, thus making possible ability to accidentally pass value of unrelated enum or just random number.
 - It induces additional developer time costs, such as price of not straightforward way of finding out the corresponding enum for concrete call, or time of sanity check for existence of developer feedback mechanism from random ints usage.
 

### Downsides of concrete implementation: 
-  No developer feedback mechanism for possible usage of unrelated to specific enum values  ([Subscribe method](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Lifecycle/LifecycleObservable.cs#L79)).
- The usage of not default [SiloLifecycleStage](https://github.com/dotnet/orleans/blob/master/src/Orleans.Runtime/Lifecycle/SiloLifecycleStage.cs#L7)  enum values increases the API exploration time by requiring to deduce if they were meant to intersect with ones from different enum, such as [GrainLifecycleStage](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core.Abstractions/Runtime/GrainLifecycleStage.cs#L15). Also comment with description of reasons behind numbers choice is missing, but it would've  been appropriate in this case.

### Alternative approach:
 
 Use statically typed parameter. Implementation: class instances instead of enums. 
### Example: 

```csharp
// Alternate entity
public interface ILifecycleStage
{
}

public interface ILifecycleObservable<in TStage> where TStage : ILifecycleStage
{
    // Alternate API
    IDisposable Subscribe(TStage stage, ILifecycleObserver observer);
}

// Implementer interface
public interface ISiloLifecycle : ILifecycleObservable<SiloLifecycleStage>
{
}

// Concrete stage definitions container
public abstract class SiloLifecycleStages 
{
    public sealed class RuntimeInitialize : ILifecycleStage
    {
    }

    public static readonly RuntimeInitialize RuntimeInitializeInstance = new RuntimeInitialize();
}
``` 
#### Callsite:
 - Current 

```csharp
 lifecycle.Subscribe(SiloLifecycleStages.RuntimeInitialize, OnRuntimeInitializeStart, ...);
```
- Alternative:
```csharp
lifecycle.Subscribe(SiloLifecycleStages.RuntimeInitializeInstance, OnRuntimeInitializeStart, ...);
```
### Benefits: 
 - Will fix the above mentioned issues.
- Stage definition class instance could either contain the optional for `Subscribe` method  `OnRuntimeInitializeStart` and `OnRuntimeInitializeStop` functions, or just `ILifecycleObserver` implementers, thus reducing signature complexity by taking `IGrainLifecycle` responsibility and removing level of indirection caused by wrapper creation in [extension method](https://github.com/dotnet/orleans/blob/master/src/Orleans.Core.Abstractions/Lifecycle/LifecycleExtensions.cs#L22) .
- Will add ease of stage definition structure functionality increase.
- Shared stages will be explicit.
 
### Downsides:
- Will add ease of stage definition structures complexity increase.
- Will bring most likely unnecessary entity into call site: new instance initialization or static one reference.

**Comment by jason-bragg** [comment]
Hello @dVakulen 
Thanks for your thoughts on this subject.

How Orleans handles lifecycles is new and is still in flux, so now is a good time to question the design.  We prototyped a couple different patterns, and, as is often the case, came up with no 'perfect' solution, so the current design reflects priorities and tradeoffs which should be questioned.

We're hoping to developed a generic staged lifecycle pattern that can work for most systems which need staged initialization and shutdown, so generalization of the pattern is important.  Our initial implementations were closer to what you proposed, with the following interface:

    public interface ILifecycleObservable<in TStage>
    {
        IDisposable Subscribe(TStage stage, ILifecycleObserver observer);
    }

See https://github.com/jason-bragg/orleans/blob/210ee60eae7c916777c194f515d684e5b4e6aa76/src/Orleans/Lifecycle/ILifecycleObservable.cs

We moved from a generic stage type to simply using 'int' for two reasons:

* Generic types limited stages to specific values, which limited the stages more than was practical.  We needed stages to be defined as general ranges, rather than fixed values to allow applications and extensions more control over order.  Consider grain lifecycle, where we've a SetupState stage defined 1000 before the Activate stage.  This allows these stages to be understood as general markers around which users can cluster their logic.  One can register logic before or after any of these stages or at various points between the stages.  This will be more important in the silo lifecycle where users may want to order initialization of logic and services that do not use grains before those that do use grains.
* Using a type was more cumbersome as all the lifecycle related logic (including logging and metrics) needed to account for the generic stage.  It also made generic implementations less useful as it tempted users to write custom logic for their generic type.

In short, we got little value out of having custom stages, it added complexity, and was more limiting.

Having said all of that, the pattern of using enums and casting them to int values may not be wise, especially since the value of the stages are defined ranges.

**Comment by dVakulen** [comment]
Thank you for the reply, @jason-bragg. 

So the choice to use `int` was based on following requirements:
 - Support ordered ranges of register logic points between stages.
 - Avoid costs of dealing with generic type. 
 - Restrict ability to increase stage definition structure functionality and complexity.

The register points ranges when implemented in such way while having great flexibility, I'd assume would be hard to use and support. It may result in magic numbers usages and hard to comprehend registered logic execution order, as it could be done from multiple places in code base, each with its own set of numbers. 

Seems like registered logic order could be represented as `LinkedList` with node being array of callbacks which can be executed in parallel; in another words - register points would depend on another points or, in case of multiple dependencies, synchronization barrier - container of required to complete before stages / points. 
Thus, suggested approach would allow to clearly define life cycle flow:

```csharp 
var grainLifecycleFlow = new List<ILifecycleParticipant>
{
    new PreSetupStateObserver(),             // ILifecycleObserver
    GrainLifecycleStage.SetupState,          // reference on stage, so that execution order would be clearly seen 
    new ParallelLifecycleParticipantsBarrier // contains participants which should be executed in parallel
    {
        new AfterSetupStateObserver(),
        new ParallelAfterSetupStateObserver(),
    },
    GrainLifecycleStage.Activate,
    new OnActivateObserver(stage, observer)
};
```

Should I also address costs of dealing with generic type and stage definition complexity considerations?

**Comment by jason-bragg** [comment]
> The register points ranges when implemented in such way while having great flexibility, I'd assume would be hard to use and support.

Possibly.  I agree that there is a tradeoff here between usability and flexibility, but I think it’s a minor one, and not terribly new.  Ring based startup has been around for quite some time.  While historically it’s been used for security and resource protection, fault tolerance and dependency management have always been part of what the pattern provides. See https://en.wikipedia.org/wiki/Protection_ring.  This has been used to start/stop logic in many posix systems for decades.  Newer patterns do exist wherein systems simply declare their dependencies and the system is smart enough to figure out the order from the dependency, but those systems are more complex.  My preference would be to start with a simple system and only graduate to more complex solutions if the simple solutions demonstrate themselves as wanting.

> Clear to reader life cycle flow definition

I agree that this is more explicit and clear regarding what the stages are, however it also introduces the following complications/costs:
* Allows for misconfiguration as components can reference stages that were not registered in the lifecycle.
* Complicates usage patterns like that which LogConsistentGrain requires, where it needs to perform an action relative to a specific stage.  See https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/LogConsistency/LogConsistentGrain.cs
* Complicates configuration as reading an object type or instance from configurations like xml or json would need be defined and documented.
* Prevents components from having defaults other then those pre-defined in the framework.

The pre-defined stages of a lifecycle (silo, grain, ..) are only markers for critical stages, the actual order systems are started is solely based off of the stages which components register for, which for the application layer should all be configured as part of the application setup.

**Comment by dVakulen** [comment]
Before starting with simple system - it would be helpful to clearly define the alternatives. 
It won't require significant effort. It can be used as reference in further discussions. It's parts could be reused for documentation.

**Comment by dVakulen** [comment]
Cycle flow definition approach discussion:
  - Misconfiguration will be prevented by means of generic constraints. 
 - Complicated usage pattern as consideration:  having alternatives to assess in written form will be useful. 
 - Complicated configuration as consideration: support multiple configuration formats as requirement. 
-  Allows flexible configuration. Consideration about preventing components from ability to override defaults - is another sign of need in clear alternatives definition.

**Comment by dVakulen** [comment]
Defining requirements in written form through means of GitHub issues would be useful:
 - Requirements: 
    - already discussed internally within the team. 
    - written in form of code. 

 - It will reduce cognitive load on external contributors.

**Comment by dVakulen** [comment]
Addressing argument from cycle flow definition approach discussion:

[LogConsistentGrain flow subscription](https://github.com/dotnet/orleans/blob/5af9c6387e1085e0464497c17db05e876116d2db/src/Orleans.Core/LogConsistency/LogConsistentGrain.cs#L54): 
 - Dependency on base class method call. 
 - Implicit dependency on `ILogConsistencyProtocolParticipant`
 - Consequences of else is not defined.
- Type erasure due to casting of  stage definition to int. 
- Implicitly defined pre and post activation callbacks (+1 ? was that reference to old-fashioned way of GitHub community to react on items?) 
- Repeated subscribe calls.

**Comment by dVakulen** [comment]
Another item in software engineering hard problems list:

 -  Finding what from set of obvious things would be appropriate to say:
    - And its more efficient to have the idea definition to be written down in form of human language before allocating energy for its implementation in machine one.

**Comment by jason-bragg** [comment]
@dVakulen,
I appreciate your input, however, there is a lot in what you're putting forward here and it seems to be going in many directions at once, which makes it difficult and time consuming to properly address all your points.  If you seek to pursue these threads, which I encourage, I suggest breaking them into specific issues so focus can be directed to specific areas.

For instance:
* Some of your comments seem related to general development process, like performing design before implementation.  Thus far Orleans team hasn't adapted much (if any) formal process, for a number of reasons, but I think we're all open to discussing this.
* Some of your comments seem related to how Orleans team interacts with the community, like posting clear requirements and design on github to include the community in the process.

The comments regarding LogConsistencyGrain seem to be a set of things objectionable about that logic, which is valuable, but not necessarily relevant to this discussion, as it was only introduced to demonstrate a use case that is needed which the proposed design did not cover.  Unless I’m misunderstanding your comments, the point being made by referencing this code was actually never addressed.  That is, I saw no update to the proposed design to account for that use case, nor an alternative method suggested, nor and argument against supporting such use cases.

--
To the original discussion, I suspect my objections to the proposal may appear as obstinance.  I apologize if I've given this impression.  I view the current design as no more than the best I've come up with in the allotted time and am very open to improvements.  My responses have been intended to help provide more transparency into the reasoning behind the choices made, not to argue that the choices were correct.

I don't see you're proposal as an alternative to the current system as much as a suggested modification, since most of the pattern is unchanged.  Since, as stated, what you're proposing is pretty much the same as the original pattern we introduced (from the api perspective) I was attempting to provide some reasons why we moved away from that pattern.

In regards to the use of enums, I've conceded that we shouldn't be using those (probably should use constants instead).  They are really just left over from the original pattern that used TStage.  In any case, they are not part of the formal abstraction, or pattern, they are simply (poor) implementation choices (by me :/) made in systems using the pattern.

Part of what I've struggled with in this system is the classic tradeoff of flexibility vs. comprehensibility.  It would be more comprehensible to have a formal set of fixed stages we go through at startup (or in the grain lifecycle) and confine users to those stages, but that would be quite limiting.  It's not clear to me that the cognitive load of using an integer with a set of pre-defined marker stages is more detrimental than limiting the flexibility available to application and extension library developers by using a fixed range.

**Comment by dVakulen** [comment]
Regarding [LogConsistencyGrain Participate method](https://github.com/dotnet/orleans/blob/5af9c6387e1085e0464497c17db05e876116d2db/src/Orleans.Core/LogConsistency/LogConsistentGrain.cs#L54) - I apologize, got distracted describing current implementation and forgot to provide the alternative.

Current: 
```csharp
public override void Participate(IGrainLifecycle lifecycle)
{
    base.Participate(lifecycle);
    lifecycle.Subscribe(GrainLifecycleStage.SetupState, OnSetupState);
    if(this is ILogConsistencyProtocolParticipant)
    {
        lifecycle.Subscribe((int)GrainLifecycleStage.Activate - 1, PreActivate);
        lifecycle.Subscribe((int)GrainLifecycleStage.Activate + 1, PostActivate);
    }
}
```

Suggested:

```csharp
public override LifecycleFlow LifecycleFlow =>
{ 
  new GrainLifecycleFlow
  {
      PreActivate,
      new StageParticipantsContainer(GrainLifecycleStage.SetupState)
      {
         OnSetupState
      },
      PostActivate
  }
} 
```
where `StageParticipantsContainer` contains logic to be executed during particular stage. Also can have inheritor prefixed with `Parallel`, which will allow to express the need in parallel execution.

**Comment by dVakulen** [comment]
Original discussion continuation: 

No, your objections to the proposal does not appear as obstinance at all. I've tried to point out that the transparency into the reasoning behind the choices made can be provided by means of dedicated issue with implementation description. 

I'm not arguing against the pattern, just attempting to provide the alternative to consider.

Could you please provide another case of flexibility limitation? 

I don't see the `fixed ranges` approach in suggested alternative. 

Thank you for your patience.

**Comment by dVakulen** [comment]
Regarding [LogConsistencyGrain Participate method](https://github.com/dotnet/orleans/blob/5af9c6387e1085e0464497c17db05e876116d2db/src/Orleans.Core/LogConsistency/LogConsistentGrain.cs#L54) - I apologize, got distracted describing current implementation and forgot to provide the alternative.

Current: 
```csharp
public override void Participate(IGrainLifecycle lifecycle)
{
    base.Participate(lifecycle);
    lifecycle.Subscribe(GrainLifecycleStage.SetupState, OnSetupState);
    if(this is ILogConsistencyProtocolParticipant)
    {
        lifecycle.Subscribe((int)GrainLifecycleStage.Activate - 1, PreActivate);
        lifecycle.Subscribe((int)GrainLifecycleStage.Activate + 1, PostActivate);
    }
}
```

Suggested:

```csharp
public override LifecycleFlow GetLifecycleFlow(IGrainLifecycle lifecycle) =>
    new LifecycleFlow
    {
        GrainLifecycleStage.Activate
            .OnBefore(PreActivate)
            .OnAfter(PostActivate),
        GrainLifecycleStage.SetupState
            .With(OnSetupState)
    };
```
where `With` implicitly will create `StageParticipantsContainer` to contain logic which are to be executed during particular stage. 
Also can have methods prefixed with `Parallel` (e.g. ParallelWith) to express the need in such kind of execution.

**Comment by jason-bragg** [comment]
Specifically looping @ReubenBond, @benjaminpetit, @xiazen, @sergeybykov in as they've all had to work with this system, likely have comments on the pattern, and may not be aware of this thread.

**Comment by dVakulen** [comment]
@jason-bragg Sorry for late reply, after https://github.com/dotnet/orleans/commit/30fcb2c60a51eea50b9dfd97e865fdd076b03c66 the issues should be mainly addressed. Remaining one with `int` stages arithmetic (The calculations such as (`GrainLifecycleStage.Activate + 1`) allows creation of implicit high coupling between subscribers) could be addressed by  defining `GrainLifecycleStage.PreActivate` wihich would spare the user of need in performing `+1` calculations, and changing int values back to enum, thus reducing flexibility. I haven't seen yet the need in such, and increasing the flexibility later in case of need won't be a breaking change, as opposed to other way change.  

  Another implementation of observable object lifecycle could be seen at [Android lifecycle](https://developer.android.com/topic/libraries/architecture/lifecycle.html). 
It uses predefined `OnStart`, `OnStop`, e.t.c. enum values without casting to `int`, while having quite similar   to the Orleans requirements.

**Comment by jason-bragg** [comment]
@dVakulen, I agree that the State+/-1 isn't great and we can add more stages as we go.  Overall we've been touching up the pattern and improving it, but it's still in flux as we vet it via usage.

Do you want to update your proposal for further consideration or are you ok with the direction of the current pattern?
